lib/embag.cc

```diff
@@ -147,138 +147,152 @@ RosBagTypes::header_t Bag::readHeader(const RosBagTypes::record_t &record) {
-        }
-
-        chunk_it->info.start_time = start_time;
-        chunk_it->info.end_time = end_time;
-        chunk_it->info.message_count = count;
-
-        break;
-      }
-      case RosBagTypes::header_t::op::UNSET:
-      default:throw std::runtime_error("Unknown record operation: " + std::to_string(uint8_t(op)));
+    connection_data.md5sum = fields->at("md5sum");
+    connection_data.message_definition = fields->at("message_definition");
+    if (fields->find("callerid") != fields->end()) {
+      connection_data.callerid = fields->at("callerid");
     }
+    if (fields->find("latching") != fields->end()) {
+      connection_data.latching = fields->at("latching") == "1";
+    }
+
+    connections_[connection_id].id = connection_id;
+    connections_[connection_id].topic = topic;
+    connections_[connection_id].data = connection_data;
+    topic_connection_map_[topic].emplace_back(&connections_[connection_id]);
+  }
+
+  /**
+   * Read the CHUNK_INFO records. These are guaranteed to be immediately after the CONNECTION records,
+   * so no need to seek the file pointer
+   */
+  for (size_t i = 0; i < chunk_count; i++) {
+    const auto chunk_info_record = readRecord(stream);
+    const auto chunk_info_header = readHeader(chunk_info_record);
+
+    RosBagTypes::chunk_info_t chunk_info;
+
+    uint32_t ver;
+    uint64_t chunk_pos;
+    RosValue::ros_time_t start_time;
+    RosValue::ros_time_t end_time;
+    uint32_t count;
+
+    chunk_info_header.getField("ver", ver);
+    chunk_info_header.getField("chunk_pos", chunk_pos);
+    chunk_info_header.getField("start_time", start_time);
+    chunk_info_header.getField("end_time", end_time);
+    chunk_info_header.getField("count", count);
+
+    chunk_info.chunk_pos = chunk_pos;
+    chunk_info.start_time = start_time;
+    chunk_info.end_time = end_time;
+    chunk_info.message_count = count;
+
+    chunk_infos_[i] = chunk_info;
+  }
+
+  /**
+   * Now that we have some chunk metadata from the CHUNK_INFO records, process the CHUNK records from
+   * earlier in the file. Each CHUNK_INFO knows the position of its corresponding chunk.
+   */
+  for (size_t i = 0; i < chunk_count; i++) {
+    const auto info = chunk_infos_[i];
+
+    // TODO: The chunk infos are not necessarily Revisit this logic if seeking back and forth across the file causes a slowdown
+    stream.seekg(info.chunk_pos, std::ios_base::beg);
+
+    const auto chunk_record = readRecord(stream);
+    const auto chunk_header = readHeader(chunk_record);
+
+    RosBagTypes::chunk_t chunk{chunk_record};
+    chunk.offset = stream.tellg();
+
+    chunk_header.getField("compression", chunk.compression);
+    chunk_header.getField("size", chunk.uncompressed_size);
+
+    if (!(chunk.compression == "lz4" || chunk.compression == "bz2" || chunk.compression == "none")) {
+      throw std::runtime_error("Unsupported compression type: " + chunk.compression);
+    }
+
+    chunk.info = info;
+
+    chunks_.emplace_back(chunk);
```

Comment:
Nit: emplace_back probably works here, but push_back might be more semantically appropriate
https://abseil.io/tips/112

Reply:
Jason has used `emplace_back` throughout the vast majority of embag, so I wanted to keep it consistent, and I felt that `emplace` didn't really cause much confusion here, as the type of `chunk` is clear.

Reply:
That sounds like consistency for the sake of consistency to me, not consistency for the sake of clarity. If you're going to continue a negative trend, I think you should have a pretty good reason to do it.

When reading this code, I had to take a little extra time to figure out why you were emplacing the object instead of pushing it back. Emplace has a specific meaning, and misusing it can (and, in my opinion, has) make it extra effort for the reader.

Don't be afraid to make improvements and improve the codebase in order to improve the direction of a codebase when there is a negative trend. Jason didn't know the difference between push_back and emplace_back when he wrote this - I know, because I asked him.

Reply:
Okay I buy that